hathor/p2p/node_sync_v2.py

```diff
@@ -0,0 +1,1188 @@
+        # being used in the sync algorithm and the code was becoming a bit too complex
+        self.send_data(tx)
+
+    @property
+    def short_peer_id(self) -> str:
+        """ Returns the id of the peer (only 7 first chars)
+        """
+        if self.protocol.peer is None or self.protocol.peer.id is None:
+            return ''
+        return self.protocol.peer.id[:7]
+
+    @property
+    def is_started(self) -> bool:
+        return self._started
+
+    def start(self) -> None:
+        """ Start sync.
+        """
+        if self._started:
+            raise Exception('NodeSyncBlock is already running')
+        self._started = True
+        self.reactor.callLater(0, self.run_sync)
+
+    def stop(self) -> None:
+        """ Stop sync.
+        """
+        if not self._started:
+            raise Exception('NodeSyncBlock is already stopped')
+        self._started = False
+        if self.call_later_id and self.call_later_id.active():
+            self.call_later_id.cancel()
+
+    def get_cmd_dict(self) -> Dict[ProtocolMessages, Callable[[str], None]]:
+        """ Return a dict of messages of the plugin.
+        """
+        return {
+            ProtocolMessages.GET_NEXT_BLOCKS: self.handle_get_next_blocks,
+            ProtocolMessages.GET_PREV_BLOCKS: self.handle_get_prev_blocks,
+            ProtocolMessages.BLOCKS: self.handle_blocks,
+            ProtocolMessages.BLOCKS_END: self.handle_blocks_end,
+            ProtocolMessages.GET_BEST_BLOCK: self.handle_get_best_block,
+            ProtocolMessages.BEST_BLOCK: self.handle_best_block,
+            ProtocolMessages.GET_BLOCK_TXS: self.handle_get_block_txs,
+            ProtocolMessages.TRANSACTION: self.handle_transaction,
+            ProtocolMessages.GET_PEER_BLOCK_HASHES: self.handle_get_peer_block_hashes,
+            ProtocolMessages.PEER_BLOCK_HASHES: self.handle_peer_block_hashes,
+            ProtocolMessages.STOP_BLOCK_STREAMING: self.handle_stop_block_streaming,
+            ProtocolMessages.GET_TIPS: self.handle_get_tips,
+            ProtocolMessages.TIPS: self.handle_tips,
+            ProtocolMessages.TIPS_END: self.handle_tips_end,
+            # XXX: overriding ReadyState.handle_error
+            ProtocolMessages.ERROR: self.handle_error,
+            ProtocolMessages.GET_DATA: self.handle_get_data,
+            ProtocolMessages.DATA: self.handle_data,
+            ProtocolMessages.RELAY: self.handle_relay,
+        }
+
+    def handle_error(self, payload: str) -> None:
+        """ Override protocols original handle_error so we can recover a sync in progress.
+        """
+        assert self.protocol.connections is not None
+        sync_checkpoints = self.protocol.connections.sync_checkpoints
+        if sync_checkpoints.is_running and sync_checkpoints.peer_syncing == self:
+            # Oops, we're syncing and we received an error, remove ourselves and let it recover
+            sync_checkpoints.peer_syncing = None
```

Comment:
There is no `sync_checkpoints.on_peer_error(self)` because it only uses one peer, I'll add a call to `sync_checkpoints.sync_error()` and leave the refactor to `sync_checkpoints.on_peer_error(self)` when implementing sync-checkpoint from multiple peers.

Reply:
I think it makes sense to refactor.

Reply:
Moved to another PR.